contribs/vsp/src/main/java/playground/vsp/pt/fare/DistanceBasedPtFareHandler.java

```diff
@@ -0,0 +1,86 @@
+package playground.vsp.pt.fare;
+
+import com.google.inject.Inject;
+import org.matsim.api.core.v01.Id;
+import org.matsim.api.core.v01.TransportMode;
+import org.matsim.api.core.v01.events.ActivityStartEvent;
+import org.matsim.api.core.v01.events.PersonArrivalEvent;
+import org.matsim.api.core.v01.events.PersonDepartureEvent;
+import org.matsim.api.core.v01.events.PersonMoneyEvent;
+import org.matsim.api.core.v01.events.handler.ActivityStartEventHandler;
+import org.matsim.api.core.v01.events.handler.PersonArrivalEventHandler;
+import org.matsim.api.core.v01.events.handler.PersonDepartureEventHandler;
+import org.matsim.api.core.v01.network.Link;
+import org.matsim.api.core.v01.network.Network;
+import org.matsim.api.core.v01.population.Person;
+import org.matsim.core.api.experimental.events.EventsManager;
+import org.matsim.core.router.StageActivityTypeIdentifier;
+import org.matsim.core.utils.geometry.CoordUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class DistanceBasedPtFareHandler implements PersonDepartureEventHandler, PersonArrivalEventHandler, ActivityStartEventHandler {
+    @Inject
+    private EventsManager events;
+
+    private final double baseFare;
+    private final double distanceFare;
+    private final Network network;
+
+    private final Map<Id<Person>, Id<Link>> personDepartureMap = new HashMap<>();
+    private final Map<Id<Person>, Id<Link>> personArrivalMap = new HashMap<>();
+
+    public DistanceBasedPtFareHandler(DistanceBasedPtFareParams params,Network network){
+        this.baseFare = params.getBaseFare();
+        this.distanceFare = params.getDistanceFare();
+        this.network = network;
+    }
+
+    @Override
+    public void handleEvent(PersonDepartureEvent event) {
+        if (event.getLegMode().equals(TransportMode.pt)){
+            if (!personDepartureMap.containsKey(event.getPersonId())){
+                personDepartureMap.put(event.getPersonId(), event.getLinkId()); // Departure place is fixed
+            }
```

Comment:
Maybe it would be nicer to use the transit stop coord instead of link.getToNode() since the stop coord can differ from the link toNode. Same for the arrival stop. This could be done using the pt interaction activities which have the stop coord.

Reply:
May I know how the PT stop is recognized in the MATSim? Is it possible to have PT stop not on the link/Node of the network?